### DISCUSSION TITLE
Duda TPI3 - Patron observer y notificaciones con java.util.TimerTask

### DISCUSSION THREAD

**Discussion by iancrespiok** [dds-jv/foro#280] [Consultas generales] [answered]
Hola como estan? Tengo una duda sobre el TPI3. Para satisfacer el requerimiento de notificar al `Usuario` cada vez que falta X tiempo para alguno de los `Evento` de su `Calendario`, lo que hice fue basicamente aplicar el patron observer donde cada `Usuario` (que aparece en `Evento` en una `List<Usuario> invitados`) seria el "observador" y cada `Evento` seria un "observable". 
Cuando el `Evento`  añade un recordatorio, notifica al `Usuario` del siguiente modo:

```
   public void anadirRecordatorio(Period tiempoAnticipacion) {
    invitados.forEach(
        invitado -> invitado.recordar(this, tiempoAnticipacion)
    );
  }
```

Y luego el `Usuario` hace `recordar()` del siguiente modo:

```
  public void recordar(Evento evento, Period tiempoAnticipacion) {
    Timer timer = new Timer();
    recordatorios.forEach(
        recordatorio -> timer.schedule(recordatorio, Date.from((evento.getInicio().minus(tiempoAnticipacion).toLocalDate()).atStartOfDay(ZoneId.systemDefault()).toInstant()))
        );
  }
```
recordatorios es: `private List<TimerTask> recordatorios;`

Y los recordatorios, como dice el enunciado pueden ser:

```
public class Mailer extends TimerTask {
  ShemailLib shemailLib;
  Usuario usuario;
  Evento evento;

  @Override
  public void run() {
    shemailLib.enviarMailA(usuario.getEmail(), "Recordatorio de " + evento.getNombre(), "HOLA");
  }
}
```
```
public class Notificador extends TimerTask {
  private Evento evento;

  @Override
  public void run() {
    System.out.println("RECORDATORIO: " +  evento.getNombre());
  }
}


```

El problema es:
En la tercer linea del metodo `recordar()` de `Usuario` hace `evento.getInicio()` y eso es un error ya que `getInicio()` es propio de `EventoUnico` pero `recordar() `deberia funcionar para la clase abstracta `Evento`, sin importar cual es. Y un poco relacionado con esto, si no pido una fecha en particular, no podria matchear con la infinidad de sucesos de un `EventoRepetitivo `para restarle el `Period tiempoAnticipacion`, seria un proceso que tardaría infinito tiempo, entonces, como hago?

La verdad que estoy bastante confundido también con si aplique bien el tema de los recordatorios. Muchas cosas tuve que buscarlas en google ya que no pude aplicar lo que vimos en clase, no se si esta bien o no, me cuesta darme cuenta si voy bien encaminado. 

Gracias

**Comment by flbulgarelli** [comment, marked as answer]
¡Hola!

Un par de cuestiones: 

1. No me termina de quedar claro dónde estaría la idea de osberver que mencionás. No veo ninguna noción de eventos en el sentido que le da el patrón, ni de observadores polimórficos. 
2. Estás utilizando planificación interna, pero como mencionamos en clase, deberías ir siempre por planificación externa cuando estamos hablando de aspectos calendarizados en períodos largos, cuya calendarización tiene que ser reobusta (en el sentido de asegurar la ejecución a toda costa) pero no necesariamente precisa (es decir, no importa tanto si la notificación arriva unos milisegundos o incluso algunos segundos tarde). 
3. El código te está quedando más extenso de lo necesario en esta línea: `        recordatorio -> timer.schedule(recordatorio, Date.from((evento.getInicio().minus(tiempoAnticipacion).toLocalDate()).atStartOfDay(ZoneId.systemDefault()).toInstant()))`. Esto se debe a que por ejemplo:    
    1  `atStartOfDay` no necesita recibir un argumento (ver https://docs.oracle.com/javase/8/docs/api/java/time/LocalDate.html#atStartOfDay--)
    2. el mensaje `minus` (https://docs.oracle.com/javase/8/docs/api/java/time/LocalDate.html#minus-long-java.time.temporal.TemporalUnit-) de `LocalDate` ya devuelve un `LocalDate` y por tanto no necesita llamar a `toLocalDate`
    3. `atStartOfDay()` devuelve un objeto que (`LocalDateTime`) que ya entiende `toInstant`
    4. Teniendo todo esto en cuenta, entiendo que debería alcanzar con  `evento.getInicio().minus(tiempoAnticipacion).atStartOfDay().toInstant()`

**Reply by flbulgarelli** [reply]
Teniendo en cuenta particularmente el punto 2, la solución al problema original que proponés probablemente sea una combinación de modificar tu diseño y de usar calendarización externa: en lugar de hacer que cada evento sepa agregar infinitos recordatorios, posibilitar que todos los eventos sepan decir (polimórficamente) si están cerca o no de su ocurrencia más cercana.

Con este comportamiento, podrás en todo momento (y periódicamente) 
   1. obtener todos los eventos del sistema que necesitan ser recordados
   2. notificar a cada uno de ellos. 

Como ves, acá no estamos generando generando timers para cada uno de los eventos, sino planificando todos los eventos conjuntamente.

**Reply by gastonprieto** [reply]
@iancrespiok

**Reply by iancrespiok** [reply]
Hola! Ya hable con Fede y me menciono que no me meta con el tema de Timers. Ya lo solucione distinto. Gracias!
